### PR TITLE
rqt_image_view: Adding the possibility to publish mouse events (updated)

### PR DESCRIPTION
**Pull request by v-lopez**
This is basically the same contents as #51 but updated to the latest code.

### CODE REVIEW

**Comment by DorianScholz** [comment]
LGTM, but I can't test it at the moment.

File: rqt_image_view/src/rqt_image_view/image_view.cpp [hunk 1]
```diff
@@ -280,6 +292,49 @@ void ImageView::saveImage()
...
+    clickLocation.y = round((double)y/(double)ui_.image_frame->height()*(double)ui_.image_frame->getImage().height());
+    clickLocation.z = 0;
+    pub_mouse_left_.publish(clickLocation);
+  }
+}
+
+void ImageView::onPubTopicChanged()
+{
+  pub_topic_custom_ = !(ui_.publish_click_location_topic_line_edit->text().isEmpty());
+  pub_mouse_left_.shutdown();
```
Comment (dirk-thomas) [inline rqt_image_view/src/rqt_image_view/image_view.cpp]: Why is this necessary? The function `onMousePublish` either create a new publisher or shuts down the previous one.

File: rqt_image_view/src/rqt_image_view/image_view.cpp [hunk 2]
```diff
@@ -77,11 +77,20 @@ void ImageView::initPlugin(qt_gui_cpp::PluginContext& context)
 
   ui_.save_as_image_push_button->setIcon(QIcon::fromTheme("image"));
   connect(ui_.save_as_image_push_button, SIGNAL(pressed()), this, SLOT(saveImage()));
+
+  pub_topic_custom_ = false;
```
Comment (dirk-thomas) [inline rqt_image_view/src/rqt_image_view/image_view.cpp]: Why do we need a boolean variable for this? The state seems to be always available in the line edit widget.

File: rqt_image_view/src/rqt_image_view/image_view.cpp [hunk 3]
```diff
@@ -280,6 +292,49 @@ void ImageView::saveImage()
...
+    std::string topicName;
+    if(pub_topic_custom_)
+    {
+      topicName = ui_.publish_click_location_topic_line_edit->text().toStdString();
+    } else {
+      if(!subscriber_.getTopic().empty())
+      {
+        topicName = subscriber_.getTopic()+"_mouse_left";
+      } else {
+        topicName = "mouse_left";
```
Comment (dirk-thomas) [inline rqt_image_view/src/rqt_image_view/image_view.cpp]: Wouldn't it be easier (and more intuitive to the user) if the line edit widget just has this value as a default? Then the user also seems the default.

File: rqt_image_view/src/rqt_image_view/image_view.cpp [hunk 4]
```diff
@@ -280,6 +292,49 @@ void ImageView::saveImage()
...
 }
 
+void ImageView::onMousePublish(bool checked)
+{
+  if(checked)
+  {
+    std::string topicName;
+    if(pub_topic_custom_)
+    {
+      topicName = ui_.publish_click_location_topic_line_edit->text().toStdString();
```
Comment (dirk-thomas) [inline rqt_image_view/src/rqt_image_view/image_view.cpp]: The topic name should be saved to and restored from the settings.

**Comment by 130s** [comment]
Sorry for closing #51 without knowing. As a way of atonement, I tested manually and it looks good; in the screencapture below, mouse cursor is at left top of the image (Ubuntu's Screenshot didn't catch cursor for some reason), which looks corresponds to the value in the topic.
rqt_image_view_clicked-point

File: rqt_image_view/src/rqt_image_view/image_view.cpp (hunk 1 above)
Comment (v-lopez) [inline rqt_image_view/src/rqt_image_view/image_view.cpp]: Agreed.

File: rqt_image_view/src/rqt_image_view/image_view.cpp (hunk 2 above)
Comment (v-lopez) [inline rqt_image_view/src/rqt_image_view/image_view.cpp]: Agreed, I just adapted the #51 code to the latest version.

I'll apply now a few fixes, I will remove entirely pub_topic_custom_ since I believe it's just confusing.

Comment (v-lopez) [inline rqt_image_view/src/rqt_image_view/image_view.cpp]: Actually I just saw why it was used. 

If you edit manually the line edit widget, the variable is set to true and what is in the line edit doesn't change even if you change topics. 
If you don't edit manually the line edit widget, when you change the subscription topic the publication topic will change accordingly.

Comment (v-lopez) [inline rqt_image_view/src/rqt_image_view/image_view.cpp]: I've left the boolean variable as I find that behaviour useful and the implementation is good enough imo.

File: rqt_image_view/src/rqt_image_view/image_view.cpp (hunk 3 above)
Comment (v-lopez) [inline rqt_image_view/src/rqt_image_view/image_view.cpp]: Check the latest commit, even if the checkbox is disabled, as soon as an image topic is selected, the publish topic will update (unless the user has modified it to a custom topic name)

File: rqt_image_view/src/rqt_image_view/image_view.cpp (hunk 4 above)
Comment (v-lopez) [inline rqt_image_view/src/rqt_image_view/image_view.cpp]: 5963a685b0b9175b6981a0c30fd48616a19832e7 saves topic name and checkbox state.

**Comment by v-lopez** [comment]
Please review again with the latest changes.

**Comment by dirk-thomas** [comment]
The PR looks good to me but I didn't test it. @ablasdel can you please try giving this a shot sometime soon and if it works merge it?

**Comment by ablasdel** [comment]
Seems to work and looks good but don't we want it against the new master branch?

**Comment by ablasdel** [comment]
@v-lopez sorry for the long delay on this. Would it be possible to squash this down to one commit?

Once it is merged into groovy-devel I'll cherry-pick it into master.

**Comment by dirk-thomas** [comment]
This shouldn't be merged into `groovy-devel`. When it is squashed you can cherry-pick it directly from the upstream repo into `master.

**Comment by v-lopez** [comment]
Squashed commits.

**Comment by ablasdel** [comment]
Thanks @v-lopez I tried to cherry-pick this into master and there were a small number of conflicts.

Since you are better versed on the changes can you open a new PR against master and make sure that this commit is properly rebased on top of master?

**Comment by v-lopez** [comment]
Sure, I'll try to do it in the next few days.

**Comment by ablasdel** [comment]
Thanks!

**Comment by ablasdel** [comment]
@v-lopez Let me know when this gets updated thanks!

**Comment by ablasdel** [comment]
closing in favor of: https://github.com/ros-visualization/rqt_common_plugins/pull/368
